### PR TITLE
Initial work for post.

### PR DESCRIPTION
**Pull request by bjorkstromm**

### CODE REVIEW

File: input/posts/2020-07-08-grpc-reflection-in-net.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+While Google's way of working with protobuf and gRPC have always been contract-first, [Marc Gravell's](https://github.com/mgravell) [protobuf-net](https://github.com/protobuf-net/protobuf-net) have always had code-first development as first priority. With Google's protobuf, you define your services and messages in .proto format and then use the protobuf compiler (protoc) to generate messages, client- and server stubs in the language of your choice. With protobuf-net, however, you can start out with C#, F# or VB.NET code. There's no need for .proto file. Code-first with protobuf-net is great if you don't need interoperability with other languages outside the .NET ecosystem, and even if you did, there have been ways to to get the contract generated in .proto format for message types. However, up until recently, if you needed the proto contract for gRPC services you were on your own. This was however fixed last weekend.
```
Comment (mgravell) [inline input/posts/2020-07-08-grpc-reflection-in-net.md]: nit: "has" always been contract-first ... "has always had code-first"

File: input/posts/2020-07-08-grpc-reflection-in-net.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+Code-first using protobuf-net.Grpc is in my opinion the simplest, and most straightforward way to port over your old WCF services to .NET Core and gRPC. It's way more simpler than the official guide over at [Microsoft Docs](https://docs.microsoft.com/en-us/dotnet/architecture/grpc-for-wcf-developers/migrate-wcf-to-grpc) which uses Google's protobuf instead of protobuf-net. With protobuf-net.Grpc, you'll most likely get to keep all your data contracts intact, while only needing to add the `Order` property to your data member attributes. Your service contracts, however, will need some rework. But in the end, these are minimal changes compared to porting your data contracts and service contracts over to .proto format. If you also need interoperability from other programming languages, it's a peace of cake to generate .proto files from your gRPC service. So, what are you waiting for? Start porting over your WCF services to .NET Core now, and make sure to use protobuf-net.Grpc when you do it.
```
Comment (mgravell) [inline input/posts/2020-07-08-grpc-reflection-in-net.md:226]: (no specific location) one other thing that may be worth mentioning: [`protogen`](https://www.nuget.org/packages/protobuf-net.Protogen/) also now has the ability to generate C# from these endpoints via the new `--grpc` option, but: it is new and slightly experimental

**Review by mgravell — COMMENTED** [review]
Looking great 👍 

Really appreciate the time taken here - and your huge input to the feature.

File: input/posts/2020-07-08-grpc-reflection-in-net.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+Tags: 
+- .NET
+- gRPC
+- Protocol Buffers
+---
+> gRPC Server Reflection is an optional extension for servers to assist clients in runtime construction of requests without having stub information precompiled into the client.
+> 
+> -- <cite>[gRPC Server Reflection Protocol](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md)</cite>
+
+The primary use case for gRPC Server Reflection is debugging and testing tools. Think cURL or Postman, but for gRPC. There's however another interesting use case, especially in the .NET world where we have code-first support through [protobuf-net.Grpc](https://github.com/protobuf-net/protobuf-net.Grpc).
```
Comment (daveaglick) [inline input/posts/2020-07-08-grpc-reflection-in-net.md]: I'd reverse the word order in the second sentence: "However there's another interesting..."

File: input/posts/2020-07-08-grpc-reflection-in-net.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+If you have been using WCF, you are probably familiar with code-first approach, and exposing metadata about the service using MEX or WSDL endpoints. These metadata endpoints could then be used to generate client stubs. Additionally, if you created a standard SOAP service, to be consumed over HTTP, interoperability with other programming languages was also possible by consuming the WSDL document. This is similar to what many of us do today with ASP.NET Core and OpenAPI. We create the controllers and models in C#, generate OpenAPI definition using [Swashbuckle](https://github.com/domaindrivendev/Swashbuckle.AspNetCore) and then generate clients in TypeScript using [Autorest](https://github.com/Azure/autorest).
```
Comment (daveaglick) [inline input/posts/2020-07-08-grpc-reflection-in-net.md]: Suggestion: "If you've been using WCF you're probably familiar with a code-first approach that exploses metadata about the service using MEX or WSDL endpoints"

File: input/posts/2020-07-08-grpc-reflection-in-net.md (hunk 1 above)
Comment (daveaglick) [inline input/posts/2020-07-08-grpc-reflection-in-net.md]: I really wish we could indicate exact words for these comments...oh well - a couple more suggestions: "to generate messages, client**,** and server stubs"..."**However, with** protobuf-net you can"..."There's no need for **a** .proto file"

File: input/posts/2020-07-08-grpc-reflection-in-net.md [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+We probably want this written to the filesystem, so we will then run the following.
+```cmd
+dotnet grpc-cli dump https://localhost:5001 Reflection.Sample.Greeter -o ./protos
+```
+
+That's it! We now have .proto files, generated from our server reflection endpoint. These .proto files can now be used to generate client stubs in the language of your choice.
+
+### Summary
+As you can see gRPC server reflection can be very useful in code-first scenarios, but it's also useful for testing and debugging. E.g. gRPC CLI supports invoking methods, and this is something I'm also planning on supporting in the .NET global tool.
```
Comment (daveaglick) [inline input/posts/2020-07-08-grpc-reflection-in-net.md]: "As you can see, gRPC server"

**Review by daveaglick — COMMENTED** [review]
Just some grammar suggestions - otherwise great article, TIL

File: input/posts/2020-07-08-grpc-reflection-in-net.md (hunk 4 above)
Comment (bjorkstromm) [reply input/posts/2020-07-08-grpc-reflection-in-net.md]: think you mean `exposes`, not  `exploses` :)

**Comment by bjorkstromm** [comment]
Thank you for the review! Post is now live here:
https://martinbjorkstrom.com/posts/2020-07-08-grpc-reflection-in-net

**Pull request merged**